README.md
# fluent-plugin-watch-objectspace

[Fluentd](https://fluentd.org/) input plugin to collect ObjectSpace information.
Additionally, plugin's process information similar to 
[fluent-plugin-watch-process](https://github.com/y-ken/fluent-plugin-watch-process).

## Installation

### RubyGems

```
$ gem install fluent-plugin-watch-objectspace
```

### Bundler

Add following line to your Gemfile:

```ruby
gem "fluent-plugin-watch-objectspace"
```

And then execute:

```
$ bundle
```

## Configuration


| parameter      | type              | description                            | default                                           |
|----------------|-------------------|----------------------------------------|---------------------------------------------------|
| watch_class    | array (optional)  | Class to be watched                    |                                                   |
| watch_interval | time (optional)   | Interval to watch object space         | `60`                                              |
| tag            | string (optional) | Tag for this input plugin              | `watch_objectspace`                               |
| modules        | array (optional)  | Modules which must be required         |                                                   |
| watch_delay    | time (optional)   | Delayed seconds until process start up | `60`                                              |
| gc_raw_data    | bool (optional)   | Collect GC::Profiler.raw_data          |                                                   |
| top_fields     | array (optional)  | Specify included fields of top command | `["VIRT", "RES", "SHR", "%CPU", "%MEM", "TIME+"]` |

### \<threshold\> section (optional) (single)

### Configuration

|parameter|type|description|default|
|---|---|---|---|
|memsize_of_all|float (optional)|Threshold rate which regards increased memsize as memory leaks|`1.3`|
|res_of_top|float (optional)|Threshold rate which regards increased RES as memory leaks||


## Usage

```
<source>
  @type watch_objectspace
  tag watch_objectspace
  modules cmetrics
  watch_class CMetrics::Counter, CMetrics::Gauge, CMetrics::Untyped
  watch_interval 60
  watch_delay 10
  <threshold>
    memsize_of_all 1.3
  </threshold>
</source>
```

If memory usage is over 1.3 times, it raise an exception.

## FAQ

### What is the difference between fluent-plugin-watch-objectspace and fluent-plugin-watch-process?

fluent-plugin-watch-process is useful cron/batch process monitoring, In contrast to it, fluent-plugin-watch-objectspace is
focused on used plugin's resource usage especially object and memory.

## Copyright

* Copyright(c) 2021- Kentaro Hayashi
* License
  * Apache License, Version 2.0


CHANGELOG.md
# 0.2.1

## Release v0.2.1 - 2021/09/30

* Fixed to use appropriate error log message
* Updated README.md about configuration section

# 0.2.0

## Release v0.2.0 - 2021/09/30

* Refactor config_section
  The threshold value must be put under <threshold> section
* Fixed an crash bug when watch_class parameter is not set

# 0.1.0 

## Release v0.1.0 - 2021/09/24

* Initial release



test/helper.rb
$LOAD_PATH.unshift(File.expand_path("../../", __FILE__))
require "test-unit"
require "test/unit/rr"
require "fluent/test"
require "fluent/test/driver/input"
require "fluent/test/helpers"

Test::Unit::TestCase.include(Fluent::Test::Helpers)
Test::Unit::TestCase.extend(Fluent::Test::Helpers)


test/plugin/test_in_watch_objectspace.rb
require "helper"
require "fluent/plugin/in_watch_objectspace"

class WatchObjectspaceInputTest < Test::Unit::TestCase
  setup do
    Fluent::Test.setup
  end

  DEFAULT_CONFIG = config_element("ROOT", "", {})

  private

  def create_driver(conf = DEFAULT_CONFIG)
    Fluent::Test::Driver::Input.new(Fluent::Plugin::WatchObjectspaceInput).configure(conf)
  end

  def create_config(params={})
    config_element("ROOT", "", params)
  end

  def default_params(data={})
    {
      "watch_delay" => 0,
      "watch_interval" => 1
    }.merge(data)
  end

  sub_test_case "configure" do
    def test_default_configuration
      d = create_driver
      assert_equal([
                     nil,
                     60,
                     60,
                     "watch_objectspace",
                     nil,
                     false
                   ],
                   [
                     d.instance.watch_class,
                     d.instance.watch_interval,
                     d.instance.watch_delay,
                     d.instance.tag,
                     d.instance.modules,
                     d.instance.gc_raw_data
                   ])
    end

    def test_customize
      config = config_element("ROOT", "", {
                                "watch_delay" => 1,
                                "watch_interval" => 2,
                                "watch_class" => ["String"],
                                "tag" => "customized",
                                "modules" => ["objspace"],
                                "gc_raw_data" => true,
                                "res_incremental_threshold_rate" => 1.1,
                                "memsize_of_all_incremental_threshold_rate" => 1.5
                              })
      d = create_driver(config)
      assert_equal([
                     ["String"],
                     2.0,
                     1.0,
                     "customized",
                     ["objspace"],
                     true
                   ],
                   [
                     d.instance.watch_class,
                     d.instance.watch_interval,
                     d.instance.watch_delay,
                     d.instance.tag,
                     d.instance.modules,
                     d.instance.gc_raw_data,
                   ])
    end

    def test_watch_class
      config = create_config(default_params({"watch_class" => ["String"]}))
      d = create_driver(config)
      d.run(expect_records: 1, timeout: 1)
      assert_equal([
                     1,
                     ["string"],
                     true
                   ],
                   [
                     d.events.size,
                     d.events.first.last["count"].keys,
                     d.events.first.last["count"]["string"] > 0
                   ])
    end

    def test_watch_interval
      config = create_config(default_params({"watch_interval" => 5}))
      d = create_driver(config)
      d.run(expect_records: 1, timeout: 5)
      assert_equal(1, d.events.size)
    end

    sub_test_case "watch_delay" do
      data(
        before_interval: [1, 0, 5],
        after_interval: [0, 10, 5]
      )
      test "watch delay" do |(count, delay, interval)|
        config = create_config(default_params({"watch_delay" => delay, "watch_interval" => interval}))
        d = create_driver(config)
        d.run(expect_records: 1, timeout: interval)
        assert_equal(count, d.events.size)
      end
    end

    sub_test_case "tag" do
      data(
        with_tag: ["changed", "changed"],
        default: ["watch_objectspace", nil]
      )
      test "tag" do |(tag, specified)|
        config = if specified
                   create_config(default_params({"tag" => specified}))
                 else
                   create_config(default_params)
                 end
        d = create_driver(config)
        d.run(expect_records: 1, timeout: 5)
        assert_equal([tag], d.events.collect { |event| event.first })
      end
    end

    sub_test_case "modules" do
      def test_invalid_module
        config = create_config(default_params({"modules" => "404",
                                               "watch_class" => "404"}))
        assert_raise do
          create_driver(config)
        end
      end

      def test_valid_module
        config = create_config(default_params({"modules" => "fluent/plugin/in_watch_objectspace",
                                               "watch_class" => "Fluent::Plugin::WatchObjectspaceInput"}))
        d = create_driver(config)
        d.run(expect_records: 1, timeout: 1)
        assert_equal([
                       ["fluent::plugin::watchobjectspaceinput"],
                       true
                     ],
                     [
                       d.events.collect { |event| event.last["count"].keys }.flatten,
                       d.events.all? { |event| event.last["count"]["fluent::plugin::watchobjectspaceinput"] > 0 }
                     ])
      end
    end

    def gc_raw_data_keys(data)
      if data.empty?
        data
      else
        data.collect do |raw_data|
          raw_data.keys
        end
      end
    end

    sub_test_case "gc_raw_data" do
      data(
        without_gc: [false, [[]]],
        with_gc: [true, [[%i(GC_FLAGS GC_TIME GC_INVOKE_TIME HEAP_USE_SIZE HEAP_TOTAL_SIZE HEAP_TOTAL_OBJECTS GC_IS_MARKED)]]]
      )
      test "gc" do |(gc, keys)|
        config = create_config(default_params({"gc_raw_data" => "true"}))
        d = create_driver(config)
        GC::Profiler.clear
        GC.start if gc
        d.run(expect_records: 1, timeout: 1)
        assert_equal([
                       1,
                       keys
                     ],
                     [
                       d.events.size,
                       d.events.collect { |event| gc_raw_data_keys(event.last["gc_raw_data"])},
                     ])
      end
    end

    sub_test_case "threshold" do
      def test_memsize_of_all
        config = config_element("ROOT", "", {
                                  "watch_delay" => 0,
                                  "watch_interval" => 2,
                                }, [config_element("threshold", "", {})])
        d = create_driver(config)
        assert_equal(1.3, d.instance.threshold.memsize_of_all)
      end

      def test_res_of_top
        config = config_element("ROOT", "", {
                                  "watch_delay" => 0,
                                  "watch_interval" => 2,
                                }, [config_element("threshold", "", {"res_of_top" => 1.5})])
        d = create_driver(config)
        assert_equal(1.5, d.instance.threshold.res_of_top)
      end
    end
  end

  sub_test_case "parser" do
    def test_top_parser
      config = config_element("ROOT", "", {
                                "watch_delay" => 0,
                                "watch_interval" => 5,
                                "watch_class" => ["String"]})
      d = create_driver(config)
      d.run(expect_records: 1, timeout:5)
      event = d.events.first
      assert_equal([1,
                    "watch_objectspace",
                    Fluent::EventTime,
                    ["pid", "count", "memory_leaks", "memsize_of_all", "virt", "res", "shr", "%cpu", "%mem", "time+"]
                   ],
                   [d.events.size,
                    event[0],
                    event[1].class,
                    event[2].keys])
    end
  end
end


lib/fluent/plugin/in_watch_objectspace.rb
#
# Copyright 2021- Kentaro Hayashi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "fluent/plugin/input"
require "fluent/config/error"
require "objspace"

module Fluent
  module Plugin
    class WatchObjectspaceInput < Fluent::Plugin::Input
      Fluent::Plugin.register_input("watch_objectspace", self)

      helpers :timer

      desc "Class to be watched"
      config_param :watch_class, :array, default: nil
      desc "Interval to watch object space"
      config_param :watch_interval, :time, default: 60
      desc "Tag for this input plugin"
      config_param :tag, :string, default: "watch_objectspace"
      desc "Modules which must be required"
      config_param :modules, :array, default: nil
      desc "Delayed seconds until process start up"
      config_param :watch_delay, :time, default: 60
      desc "Collect GC::Profiler.raw_data"
      config_param :gc_raw_data, :bool, default: false
      desc "Specify included fields of top command"
      config_param :top_fields, :array, default: ["VIRT", "RES", "SHR", "%CPU", "%MEM", "TIME+"]

      config_section :threshold, required: false, multi: false do
        desc "Threshold rate which regards increased memsize as memory leaks"
        config_param :memsize_of_all, :float, default: 1.3
        desc "Threshold rate which regards increased RES as memory leaks"
        config_param :res_of_top, :float, default: nil
      end

      def configure(conf)
        super(conf)
        if @modules
          @modules.each do |mod|
            begin
              require mod
            rescue LoadError
              raise Fluent::ConfigError.new("BUG: module <#{mod}> can't be loaded")
            end
          end
        end
        @warmup_time = Time.now + @watch_delay
        @source = {}
        GC::Profiler.enable
      end

      def start
        super
        timer_execute(:execute_watch_objectspace, @watch_interval, &method(:refresh_watchers))
      end

      def parse_top_result(content)
        record = {}
        fields = content.split("\n")[-2].split
        values = content.split("\n").last.split
        fields.each_with_index do |field, index|
          next unless @top_fields.include?(field)
          case field
          when "USER", "S", "TIME+", "COMMAND"
            record[field.downcase] = values[index]
          when "PID", "PR", "NI", "VIRT", "RES", "SHR"
            record[field.downcase] = values[index].to_i
          else
            record[field.downcase] = values[index].to_f
          end
        end
        record
      end

      def refresh_watchers
        return if Time.now < @warmup_time
        
        pid = Process.pid
        record = {
          "pid" => pid,
          "count" => {},
          "memory_leaks" => false,
          "memsize_of_all" => ObjectSpace.memsize_of_all
        }

        begin
          content = IO.popen("top -p #{pid} -b -n 1") do |io|
            io.read
          end
          record.merge!(parse_top_result(content))

          if @gc_raw_data
            record["gc_raw_data"] = GC::Profiler.raw_data
          end
          if @watch_class
            @watch_class.each do |klass|
              record["count"]["#{klass.downcase}"] = ObjectSpace.each_object(Object.const_get(klass)) { |x| x }
            end
          end

          if @source.empty?
            record["memsize_of_all"] = ObjectSpace.memsize_of_all
            @source = record
          end

          check_threshold(record)
          es = OneEventStream.new(Fluent::EventTime.now, record)
          router.emit_stream(@tag, es)
        rescue => e
          $log.error(e.message)
        end
      end

      def check_threshold(record)
        return unless @threshold

        if @threshold.res_of_top
          if @source["res"] * @threshold.res_of_top < record["res"]
            record["memory_leaks"] = true
            message = sprintf("Memory usage is over than expected, threshold res_of_top rate <%f>: %f > %f * %f",
                              @threshold.res_of_top, record["res"],
                              @source["res"], @threshold.res_of_top)
            raise message
          end
        end
        if @threshold.memsize_of_all
          if @source["memsize_of_all"] * @threshold.memsize_of_all < record["memsize_of_all"]
            record["memory_leaks"] = true
            message = sprintf("Memory usage is over than expected, threshold of memsize_of_all rate <%f>: %f > %f * %f",
                              @threshold.memsize_of_all, record["memsize_of_all"],
                              @source["memsize_of_all"], @threshold.memsize_of_all)
            raise message
          end
        end
      end

      def shutdown
      end
    end
  end
end
